### PR TITLE
fix spellcheck for editor2

### PR DESCRIPTION
**Pull request by petrjasek**
test a different rendering strategy, instead of updating the text
with markup for hiliting create a clone, do the hiliting there and
render it below the actual text. advantage is that there is no
cursor manipulation involved, it doesn't change markup so it doesn't
need to store and restore cursor position.

known issues:
- find and replace - before it was using hilite node to replace the
  word, it will require some changes to manipulate the text instead
- corrections dropdown - because the errors are rendered below the text,
  there would be no click event on this.

### CODE REVIEW

File: scripts/superdesk/editor2/editor.js [hunk 1]
```diff
@@ -325,19 +325,26 @@ function EditorService(spellcheck, $rootScope, $timeout, $q, _, renditionsServic
...
-        hiliteToken(node, token, className);
-        $timeout(function() {
-            hilite(node, tokens, className, true);
-        }, 0, false);
+        // remove old hilite nodes if any
+        var clones = parentNode.getElementsByClassName(CLONE_CLASS);
+        Array.prototype.forEach.call(clones, function(clone) {
+            parentNode.removeChild(clone);
+        });
+
```
Comment (naomiaro) [inline scripts/superdesk/editor2/editor.js]: getElementsByClassName returns a live collection, would be better to use slice to make an actual array from this DOM  collection first.

File: scripts/superdesk/editor2/editor.js [hunk 2]
```diff
@@ -1113,7 +1119,26 @@ angular.module('superdesk.editor2', [
...
+                     * Test if given point {x, y} is in given bouding rectangle.
+                     */
+                    function isPointInRect(point, rect) {
+                        return rect.left < point.x && rect.right > point.x && rect.top < point.y && rect.bottom > point.y;
+                    }
+
                     editorElem.on('contextmenu', function(event) {
+                        var err, pos;
+                        var point = {x: event.clientX, y: event.clientY};
+                        var errors = elem[0].parentNode.getElementsByClassName('sderror');
```
Comment (naomiaro) [inline scripts/superdesk/editor2/editor.js]: This is also a live array, the length can't really be trusted/much slower to use than using slice to create a real array.

File: scripts/superdesk/editor2/editor.js [hunk 3]
```diff
@@ -1113,7 +1119,26 @@ angular.module('superdesk.editor2', [
...
+                            pos = err.getBoundingClientRect();
+                            if (isPointInRect(point, pos)) {
+                                event.preventDefault();
+                                return;
+                            }
+                        }
+
                         if (editor.isErrorNode(event.target)) {
                             event.preventDefault();
                             var menu = elem[0].getElementsByClassName('dropdown-menu')[0],
```
Comment (naomiaro) [inline scripts/superdesk/editor2/editor.js]: hmm this is an old one but still...

**Comment by petrjasek** [comment]
@naomiaro using slice was slower than using htmlcollection, was this the concern or something else?

**Comment by ioanpocol** [comment]
:+1:

**Comment by naomiaro** [comment]
@petrjasek concern was mostly it's a live collection, so the length is not fixed. It will update when you remove items so it's not the safest to use them in for loops an add/delete items from the collection.

**Comment by petrjasek** [comment]
I see thx for info

**Comment by naomiaro** [comment]
:+1:

**Pull request merged**